### PR TITLE
HomeworkFive

### PR DESCRIPTION
**Pull request by YauheniMakarenko**

### CODE REVIEW

File: src/homeworkFive/ShopInterface.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package homeworkFive;
+
+import java.util.List;
+import java.util.Map;
+
+public interface ShopInterface {
+    void addProductShop(Product product);
+    Map getCheck(List<Integer> list);
```
Comment (vitalyahacker) [inline src/homeworkFive/ShopInterface.java]: check is not map
check is object which contais map and total price

File: src/homeworkFive/ShopOne.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import java.util.Map;
+
+public class ShopOne implements ShopInterface {
+    private List<Product> productList = new ArrayList<>();
+
+    public void addProductShop(Product product) {
+        productList.add(product);
+    }
+
+    public Map<String, Integer> getCheck(List<Integer> list) {
```
Comment (vitalyahacker) [inline src/homeworkFive/ShopOne.java]: check is not map

File: src/homeworkFive/ShopTwo.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
+package homeworkFive;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class ShopTwo implements ShopInterface {
+    private Map<Product, Integer> mapProduct = new HashMap<>();
```
Comment (vitalyahacker) [inline src/homeworkFive/ShopTwo.java]: you cannot use product as key in hash map
guess why

File: src/homeworkFive/ShopTwo.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        if (!mapProduct.containsKey(product)) {
+            mapProduct.put(product, 1);
+        } else {
+            mapProduct.put(product, mapProduct.get(product) + 1);
+        }
+    }
+
+    public Map<String, Integer> getCheck(List<Integer> list) {
+        int count;
+        for (Map.Entry<Product, Integer> p : mapProduct.entrySet()) {
```
Comment (vitalyahacker) [inline src/homeworkFive/ShopTwo.java]: very bad solution
map gives you ability to get element for O(1) complexity
but you make it work for O(n) complexity
it's very bad

File: src/homeworkFive/Product.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
+package homeworkFive;
+
+public abstract class Product {
+    private String name;
+    private int id;
+    private double price;
```
Comment (v-bliakharchu-idf) [inline src/homeworkFive/Product.java]: make class immutable
final fields and no getters

**Pull request merged**